rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+
+However, while the above syntax would be _new_ and thus not a breaking change,
+for existing code, adding lexical scoping _would_ be a breaking change:
+
+```coffee
+foo = "bar"
+
+{
+  foo = "baz"
+}
+
+foo
+```
+
+Previously, `foo` would return `"baz"` when the program runs, but with lexical
+scoping, it would be set to `"bar"` instead (depending on if we also added
+let-bindings or not).
+
+Because of this breaking change, this RFC proposes we add lexical scoping
+selectively, function-closures _will_ have _partial_ lexical scoping, anything else
+_won't_.
+
+What this means for function-closures is best demonstrated as an example:
+
+```coffee
+value1 = 1
+
+map(.) |key, value2| {
+  value1 = 1.1
+  value3 = 3
+
+  [key, value2]
+}
+
+value1
+value2
+value3
+```
+
+In the above example, `value1` returns `1.1`, `value2` returns an "undefined
+variable" error at compile-time, and `value3` returns `3`. In other words, lexical-scoping
+_only_ applies to the provided closure variables, _not_ to any variables defined
+inside the closure itself.
+
+This prevents people from accidentally overwriting previous variables, unless
+they explicitly do so within the closure-block, which is how it currently works
+in regular blocks as well.
+
+While there is _some_ inconsistency here, the inconsistency is well-isolated, it
+prevents us from having to introduce a breaking change right now, and it also
+still leaves open the option of adding lexical scoping before a 1.0 release of
+VRL.
+
+#### Returning Two-Element Array
+
+We require the `map` function closure to return a two-element `array` type.
+Without this requirement, mapping would work as follows:
+
+```coffee
+map(.) |key, _| {
+  key = upcase(key)
+}
+```
+
+That is, `key` would be a "special variable" inside the closure, which modifies
+the actual key of the record within the object.
+
+This doesn't fit existing patterns in VRL. It looks as if there's a _dangling_
+variable `key` at the end that remains unused, but because we special-cased this
+situation, it would instead magically update the actual key in the object after
+the closure runs to completion.
+
+This can become more difficult to reason about if/when we introduce control-flow
+statements such as `break`, as you could have set `key` before calling `break`,
+which would then either still mutate the actual key, or not, depending on how we
+implement `break`, but either way, the program itself becomes less readable, and
+operators have to read the language documentation to understand the semantic
+differences between how code behaves _inside_ a function-closure and _outside_.
+
+Instead, the `map` function-closure is required to return a two-element array of
+`[key, value]`, which the function machinery then uses to update the actual
+values of the object record, e.g.:
+
+```coffee
+map(.) |key, value| {
+  key = upcase(key)
+
+  # The array return-value clearly defines the eventual key and value values.
+  [key, value]
+}
+```
```

Comment:
I slightly prefer being more explicit and returning an object rather than an array here:

```coffee
{ key = key, value = value }
```

Reply:
Being more explicit aligns with the VRL design doc - perhaps treated as function arguments today?

```coffee
{ key = key, value = value }
{ key, value }
{ value = value, key = key }
```

Would all be equivalent.

Reply:
I'm torn on this. I somewhat think the required object is _too_ verbose. The second example that @spencergilbert mentions is something @StephenWakely and I discussed offline as well, but I'm not convinced we can make that work, due to parsing ambiguity (`{ foo }` can be both a block expression, or a keyless-object expression).

Rust has the same issue with `(foo)` being either a grouped-expression or a tuple-expression. They solve this by requiring you to add an extra `,` for single-element tuple expressions: `(foo,)`. We could do something similar.

Reply:
That's fair - it's not a hill I'm interested to die on 😄 

```coffee
[key, value]
[value, key]
```

Isn't  all that different 😉

Reply:
As noted in Slack, I'm leaning towards moving forward as-is, given that returning an array or object is based on the `map` function implementation, and doesn't need any compiler changes. Additionally, actually building that map function will be (1) the last piece of the implementation puzzle, and (2) relatively easy compared to all the other work, so we don't have to sweat this detail as much as others before we start the implementation work.

Reply:
Yeah, I agree, I think returning an object is a bit verbose. It would be nice to have real tuples, but I think the two-element array is next best.